Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/Constraints/TransformConstraint.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Utilities;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI
+{
+    /// <summary>
+    /// Base class for all constraints
+    /// </summary>
+    public abstract class TransformConstraint : MonoBehaviour
+    {
+        #region Properties
+
+        [SerializeField]
+        [Tooltip("Transform being constrained. Defaults to the object of the component.")]
+        private Transform targetTransform = null;
+
+        /// <summary>
+        /// Transform that we intend to apply constraints to
+        /// </summary>
+        public Transform TargetTransform
+        {
+            get => targetTransform;
+            set => targetTransform = value;
+        }
+
+        protected MixedRealityPose worldPoseOnManipulationStart;
+
+        public abstract TransformFlags ConstraintType { get; }
+
+        #endregion Properties
+
+        #region MonoBehaviour Methods
+
+        public virtual void Start()
+        {
+            if (TargetTransform == null)
+            {
+                TargetTransform = transform;
+            }
+        }
+
+        #endregion MonoBehaviour Methods
+
+        #region Public Methods
+
+        /// <summary>
+        /// Intended to be called on manipulation started
+        /// </summary>
+        public virtual void Initialize(MixedRealityPose worldPose)
+        {
+            worldPoseOnManipulationStart = worldPose;
+        }
+
+        /// <summary>
+        /// Abstract method for applying constraints to transforms during manipulation
+        /// </summary>
+        public abstract void ApplyConstraint(ref MixedRealityTransform transform);
```

Comment:
How is this different from a solver?

Reply:
This only constrains in what way we can manipulate an object, whereas if I understand it, a solver continuously updates an object's transforms in some way relative to another transform.

For example, if we had a "fixed distance from head" constraint, that limits a transforms position so that while manipulating, the object remains at a fixed distance from the head, but once we stop manipulating we can move away from the object and it will stay in place.

Whereas, if we had a "fixed distance from head" solver for example, that would continuously update the position of the object. When we move, the object will always move with us, remaining at a fixed distance from the head. 

The solver also needs to make some decisions as to where to move the object to, whereas with constraints this is managed by manipulation.